pkg/api/types.go

```diff
@@ -450,10 +452,81 @@ type StepConfiguration struct {
 // if no explicit output tag is provided, the name
 // of the image is used as the tag.
 type InputImageTagStepConfiguration struct {
+	InputImage `json:",inline"`
+	Sources    *[]ImageStreamSource `json:"-"`
+}
+
+func (config InputImageTagStepConfiguration) Matches(other InputImage) bool {
+	return config.InputImage == other
+}
+
+func (config InputImageTagStepConfiguration) FormattedSources() string {
+	var formattedSources string
+	tests := sets.String{}
+	if config.Sources != nil && len(*config.Sources) > 0 {
+		for _, source := range *config.Sources {
+			switch source.SourceType {
+			case ImageStreamSourceTest:
+				tests.Insert(source.Name)
+			default:
+				formattedSources += fmt.Sprintf("%s,", string(source.SourceType))
+			}
+		}
+
+		if len(tests) > 0 {
+			var testNames []string
+			for k := range tests {
+				testNames = append(testNames, k)
+			}
+			sort.Strings(testNames)
+			formattedSources += fmt.Sprintf("test steps: %s", strings.Join(testNames, ","))
+		}
+
+		return strings.TrimSuffix(formattedSources, ",")
+	} else {
+		return ""
+	}
+}
+
+func (config *InputImageTagStepConfiguration) AddSources(sources ...ImageStreamSource) *InputImageTagStepConfiguration {
+	if config.Sources == nil {
+		config.Sources = &sources
+	} else {
+		*config.Sources = append(*config.Sources, sources...)
+	}
+	return config
+}
+
+type InputImage struct {
 	BaseImage ImageStreamTagReference         `json:"base_image"`
 	To        PipelineImageStreamTagReference `json:"to,omitempty"`
 }
 
+type ImageStreamSourceType string
+
+const (
+	ImageStreamSourceRoot    ImageStreamSourceType = "root"
+	ImageStreamSourceBase    ImageStreamSourceType = "base_image"
+	ImageStreamSourceBaseRpm ImageStreamSourceType = "base_rpm_image"
+	ImageStreamSourceTest    ImageStreamSourceType = "test step"
+)
+
+type ImageStreamSource struct {
+	SourceType ImageStreamSourceType
+	Name       string
+}
+
+func NewImageStreamSource(sourceType ImageStreamSourceType, source *string) ImageStreamSource {
```

Comment:
What's the utility of passing in a pointer for `source` here? If we don't pass that in, can users just create literals instead of calling this?